File: src/main/java/com/jcloisterzone/action/CastleAction.java
package com.jcloisterzone.action;

import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.game.capability.CastleCapability.CastleToken;
import com.jcloisterzone.wsio.message.PlaceTokenMessage;
import com.jcloisterzone.wsio.message.WsInGameMessage;
import io.vavr.collection.Set;

public class CastleAction extends AbstractPlayerAction<FeaturePointer> implements SelectFeatureAction {

    public CastleAction(Set<FeaturePointer> options) {
        super(options);
    }

    @Override
    public WsInGameMessage select(FeaturePointer ptr) {
        return new PlaceTokenMessage(CastleToken.CASTLE, ptr);
    }

    @Override
    public String toString() {
        return "place castle";
    }

}

File: src/main/ts/com/jcloisterzone/action/CastleAction.ts
import type { Set } from "../../../io/vavr/Set.js";
import type { FeaturePointer } from "../board/pointer/FeaturePointer.js";
import { AbstractPlayerAction } from "./AbstractPlayerAction.js";
import type { SelectFeatureAction } from "./SelectFeatureAction.js";

/** Offer to convert a just-completed-eligible 2-tile city into a castle. */
export class CastleAction extends AbstractPlayerAction<FeaturePointer> implements SelectFeatureAction {
  static readonly simpleName = "CastleAction";

  constructor(options: Set<FeaturePointer>) {
    super(options);
  }
}
